File: BKJN_CT/SmallLighter/1012.js
const filePath =
  process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt';
const input = require('fs')
  .readFileSync(filePath)
  .toString()
  .trim()
  .split('\n');

const t = Number(input.shift());

const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

let graph;
let M, N, K;

const bfs = (x, y) => {
  const q = [[x, y]];
  while (q.length) {
    const [x, y] = q.shift();

    if (!graph[x][y]) continue;
    else graph[x][y] = 0;

    for (let i = 0; i < 4; i++) {
      const nx = x + dx[i];
      const ny = y + dy[i];

      if (nx >= 0 && nx < M && ny >= 0 && ny < N && graph[nx][ny]) {
        q.push([nx, ny]);
      }
    }
  }
};

for (let i = 0; i < t; i++) {
  [M, N, K] = input.shift().split(' ').map(Number);
  graph = [];
  let cnt = 0;

  for (let j = 0; j < M; j++) {
    graph.push(new Array(N).fill(0));
  }

  for (let k = 0; k < K; k++) {
    const [x, y] = input.shift().split(' ').map(Number);
    graph[x][y] = 1;
  }

  for (let l = 0; l < M; l++) {
    for (let m = 0; m < N; m++) {
      if (graph[l][m]) {
        bfs(l, m);
        cnt++;
      }
    }
  }
  console.log(cnt);
}

File: BKJN_CT/SmallLighter/1012.py
import sys
from collections import deque

sys.stdin = open("./BKJN_CT/SmallLighter/input.txt", "r")

input = sys.stdin.readline

tc = int(input())

dx = [-1,1,0,0]
dy = [0,0,-1,1]

def bfs(x,y):
    q = deque([[x,y]])
    while(q):
        x,y = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < M and 0 <= ny < N and graph[nx][ny] == 1:
                graph[nx][ny] = 0
                q.append([nx,ny])

for _ in range(tc):
    M, N, K = map(int, input().split())
    graph = [[0] * (N) for _ in range(M)]
    cnt = 0

    for _ in range(K):
        [x, y] = map(int, input().split())
        graph[x][y] = 1
    
    for i in range(M):
        for j in range(N):
            if graph[i][j] == 1:
                bfs(i,j)
                cnt += 1
    
    print(cnt)